### PR TITLE
feat(payments): add charge/refund tracking for new subscriptions

### PR DESCRIPTION
**Pull request by hdahlheim**
- [x] tested Localy
- [x] tested on staging

### CODE REVIEW

**Review by lucianak — COMMENTED** [review]
looks good, just added some smaller comments. Finance confirmed that sometimes we do issue partial refunds but I think we don't need to have a separate status for it

File: packages/backend-modules/payments/migrations/sql/20240923154325-payments-charges-down.sql [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    "externalId" text NOT NULL,
+    "createdAt" timestamptz DEFAULT now(),
+    "updatedAt" timestamptz DEFAULT now(),
+    "paidAt" timestamptz,
+    "refundedAt" timestamptz,
+    CONSTRAINT fk_charge_user_id FOREIGN KEY("userId")
+       REFERENCES public.users("id"),
+    CONSTRAINT fk_charge_invoice_id FOREIGN KEY("invoiceId")
+       REFERENCES payments.invoices("id")
+)
```
[the remark this answers is not in the document]
Comment (hdahlheim) [reply packages/backend-modules/payments/migrations/sql/20240923154325-payments-charges-down.sql:31]: true that was an error during creation. 🤣

File: packages/backend-modules/payments/migrations/sql/20240923154325-payments-charges-up.sql [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    'TWINT',
+    'PAYPAL'
+);
+
+CREATE TABLE IF NOT EXISTS payments."charges" (
+    "id" uuid default uuid_generate_v4() PRIMARY KEY,
+    "company" payments.company NOT NULL,
+    "provider" payments.payment_provider default 'STRIPE',
+    "externalId" text NOT NULL,
+    "invoiceId" uuid, -- invoice that is being charged
```
[the remark this answers is not in the document]
Comment (hdahlheim) [reply packages/backend-modules/payments/migrations/sql/20240923154325-payments-charges-up.sql:20]: If we start tracking charges for cards or something we might not have an invoice. Also if we start deleting invoices related to users we might be missing the invoice.

File: packages/backend-modules/payments/lib/handlers/stripe/invoicePaymentSucceded.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+export function mapChargeArgs(
+  company: Company,
+  invoiceId: string,
+  charge: Stripe.Charge,
+) {
+  let paymentMethodType: string | null = null
+  if (charge?.payment_method_details?.card) {
+    paymentMethodType = 'CARD'
```
[the remark this answers is not in the document]
Comment (hdahlheim) [reply packages/backend-modules/payments/lib/handlers/stripe/invoicePaymentSucceded.ts:49]: The provider is saved in the row and if not explicitly set it defaults to stripe. Adding it to the payment method type in addition to that is not necessary I my opinion.

File: packages/backend-modules/payments/lib/payments.ts [hunk 4]
```diff
@@ -29,6 +29,7 @@ import {
 } from './transactionals/sendTransactionalMails'
 import { enforceSubscriptions } from '@orbiting/backend-modules-mailchimp'
 import { getConfig } from './config'
+// eslint-disable-next-line @typescript-eslint/no-require-imports
```
[the remark this answers is not in the document]
Comment (hdahlheim) [reply packages/backend-modules/payments/lib/payments.ts:32]: Not without changing some typescript settings. Destucturing using import here leads to the following error:

`This module can only be referenced with ECMAScript imports/exports by turning on the 'allowSyntheticDefaultImports' flag and referencing its default export.`

the alternative would be requiring it like this:

```
import Auth = require('@orbiting/backend-modules-auth')
const UserEvents = Auth.UserEvents
```

or using the import syntax:

```
import Auth from '@orbiting/backend-modules-auth'
const UserEvents = Auth.UserEvents
```

File: packages/backend-modules/payments/lib/handlers/stripe/invoicePaymentSucceded.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (lucianak) [reply packages/backend-modules/payments/lib/handlers/stripe/invoicePaymentSucceded.ts:49]: sorry, I missed that column somehow 🙈

**Comment by hdahlheim** [comment]
> looks good, just added some smaller comments. Finance confirmed that sometimes we do issue partial refunds but I think we don't need to have a separate status for it

As far as I was able to tell we also did not have a status for partial refunds in the past.

**Pull request merged**